app/src/org/commcare/dalvik/geo/HereFunctionHandler.java

```diff
@@ -0,0 +1,174 @@
+package org.commcare.dalvik.geo;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import org.commcare.dalvik.activities.EntitySelectActivity;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.condition.IFunctionHandler;
+import org.javarosa.core.model.data.GeoPointData;
+import org.odk.collect.android.utilities.GeoUtils;
+
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * Allows evaluation contexts to support the here() XPath function,
+ * which returns the current location.
+ *
+ * In addition, an EntitySelectActivity can register itself to be refreshed whenever
+ * a new value of here() is obtained (whenever the location changes).
+ *
+ * No locations are requested if here() is never evaluated.
+ *
+ * @author Forest Tong
+ */
+public class HereFunctionHandler implements IFunctionHandler, LocationListener {
+    public static final String HERE_NAME = "here";
+    private GeoPointData location;
+
+    private LocationManager mLocationManager;
+
+    private boolean requestingLocationUpdates;
+    private boolean locationGoodEnough;
+
+    private final Context context = CommCareApplication._().getApplicationContext();
+
+    // If there are more general uses for HereFunctionHandler, the type of this field can be
+    // generalized to a listener interface.
+    private EntitySelectActivity entitySelectActivity;
+
+    public HereFunctionHandler() {}
+
+    public void registerEntitySelectActivity(EntitySelectActivity entitySelectActivity) {
+        this.entitySelectActivity = entitySelectActivity;
+    }
+
+    public void unregisterEntitySelectActivity() {
+        this.entitySelectActivity = null;
+    }
+
+    // The EntitySelectActivity must subscribe before this method is called if a fresh location is desired.
+    public Object eval(Object[] args, EvaluationContext ec) {
+        if (entitySelectActivity != null && !entitySelectActivity.containsHereFunction) {
+            refreshLocation();
+            allowGpsUse();
+            entitySelectActivity.containsHereFunction = true;
```

Comment:
It is generally bad form in Java land to modify the fields of another class directly. You should do this via getters and setters. While this introduces boilerplate, it makes for more extensible code: maybe you want to synchronize your setters across threads, maybe you want to cache computations and invalidate them when a setter is called, etc.
Most importantly, this is convention: readers of code will not expect for class field mutations to occur outside of that class, and not look for it when they are trying to understand the code.